File: packages/agent-rust/src/edge.rs
//! Edge read-path helpers. Mirrors the pure functions in `packages/shared/src/edge.ts`
//! that the edge/`both` agent uses to turn app-published upstream shards into routes.

use std::collections::BTreeMap;

use crate::types::UpstreamShard;

/// One upstream target an edge routes to (an app replica reachable over the VPC).
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct EdgeBackend {
    pub domain: String,
    pub private_ip: String,
    pub host_port: i64,
}

/// Build Caddy upstreams from app-published routing shards (edge agent read path),
/// grouped per domain in first-seen order.
pub fn build_edge_backends_from_shards(shards: &[UpstreamShard]) -> Vec<(String, Vec<EdgeBackend>)> {
    let mut order: Vec<String> = Vec::new();
    let mut by_domain: BTreeMap<String, Vec<EdgeBackend>> = BTreeMap::new();
    for shard in shards {
        for b in &shard.backends {
            if !by_domain.contains_key(&b.domain) {
                order.push(b.domain.clone());
            }
            by_domain.entry(b.domain.clone()).or_default().push(EdgeBackend {
                domain: b.domain.clone(),
                private_ip: shard.private_ip.clone(),
                host_port: b.host_port,
            });
        }
    }
    order
        .into_iter()
        .map(|d| {
            let backends = by_domain.remove(&d).expect("domain was inserted");
            (d, backends)
        })
        .collect()
}

/// First health path published for a domain across shards (for Caddy active checks).
pub fn edge_health_path_for_domain(shards: &[UpstreamShard], domain: &str) -> Option<String> {
    for shard in shards {
        for b in &shard.backends {
            if b.domain == domain {
                if let Some(path) = &b.health_path {
                    return Some(path.clone());
                }
            }
        }
    }
    None
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::types::UpstreamBackend;

    fn shard(node_id: &str, ip: &str, backends: Vec<UpstreamBackend>) -> UpstreamShard {
        UpstreamShard {
            node_id: node_id.into(),
            private_ip: ip.into(),
            updated_at: "t".into(),
            backends,
        }
    }

    fn backend(domain: &str, host_port: i64, health_path: Option<&str>) -> UpstreamBackend {
        UpstreamBackend {
            domain: domain.into(),
            host_port,
            health_path: health_path.map(|s| s.to_string()),
        }
    }

    #[test]
    fn groups_backends_across_shards_by_domain() {
        let shards = vec![
            shard("app-1", "10.0.1.5", vec![backend("a.com", 20001, Some("/h"))]),
            shard("app-2", "10.0.1.6", vec![backend("a.com", 20001, None)]),
        ];
        let grouped = build_edge_backends_from_shards(&shards);
        assert_eq!(grouped.len(), 1);
        let (domain, backends) = &grouped[0];
        assert_eq!(domain, "a.com");
        assert_eq!(
            backends,
            &vec![
                EdgeBackend {
                    domain: "a.com".into(),
                    private_ip: "10.0.1.5".into(),
                    host_port: 20001
                },
                EdgeBackend {
                    domain: "a.com".into(),
                    private_ip: "10.0.1.6".into(),
                    host_port: 20001
                },
            ]
        );
    }

    #[test]
    fn finds_the_first_health_path_for_a_domain() {
        let shards = vec![
            shard("app-1", "10.0.1.5", vec![backend("a.com", 20001, None)]),
            shard("app-2", "10.0.1.6", vec![backend("a.com", 20001, Some("/health"))]),
        ];
        assert_eq!(
            edge_health_path_for_domain(&shards, "a.com").as_deref(),
            Some("/health")
        );
        assert_eq!(edge_health_path_for_domain(&shards, "missing.com"), None);
    }
}

File: packages/shared/src/edge.ts
import { z } from "zod";
import { type DesiredState, serviceKey } from "./desired";
import type { NodeRegistryEntry } from "./registry";
import type { NodeStatus } from "./status";

/** Advisory `edge.json`: the domains an edge node fronts (for cold-start + display). */
export const EdgeConfigSchema = z
  .object({
    nodeId: z.string().min(1),
    domains: z.array(z.string().min(1)).default([]),
    updatedAt: z.string(),
  })
  .strict();
export type EdgeConfig = z.infer<typeof EdgeConfigSchema>;

export interface EdgeBackend {
  domain: string;
  privateIp: string;
  hostPort: number;
}

/** One upstream target an app agent publishes for its edge. */
export const UpstreamBackendSchema = z
  .object({
    domain: z.string().min(1),
    hostPort: z.number().int().positive(),
    healthPath: z.string().optional(),
  })
  .strict();
export type UpstreamBackend = z.infer<typeof UpstreamBackendSchema>;

/** Routing telemetry an app agent writes into an edge node's upstream prefix. */
export const UpstreamShardSchema = z
  .object({
    nodeId: z.string().min(1),
    privateIp: z.string().min(1),
    updatedAt: z.string(),
    backends: z.array(UpstreamBackendSchema).default([]),
  })
  .strict();
export type UpstreamShard = z.infer<typeof UpstreamShardSchema>;

export function parseUpstreamShard(input: unknown): UpstreamShard {
  return UpstreamShardSchema.parse(input);
}

/** One cluster member joined with its live S3 state, as seen by the edge. */
export interface ClusterNode {
  entry: NodeRegistryEntry;
  status: NodeStatus | null;
  desired: DesiredState | null;
}

/**
 * Build the per-domain upstream list by CROSS-READING other nodes' desired.json +
 * status.json: for every app node in the cluster, join its desired services
 * (domain + edge ownership) with its status (running, healthy replicas + host
 * ports), keeping only services routed by `edgeNodeId`.
 *
 * ⚠️ This is NOT the production routing path and intentionally VIOLATES the
 * "push-based routing, never cross-reads" invariant (see CLAUDE.md). The live
 * edge agent uses `buildEdgeBackendsFromShards` (below), which reads only its own
 * `upstream/*` prefix. This function is retained for tests / debugging that need
 * to compute the expected routing table from raw node state — do not call it from
 * the agent.
 */
export function buildEdgeBackendsByCrossRead(
  edgeNodeId: string,
  nodes: ClusterNode[],
): Map<string, EdgeBackend[]> {
  const result = new Map<string, EdgeBackend[]>();

  for (const node of nodes) {
    if (node.entry.nodeId === edgeNodeId) continue;
    const privateIp = node.entry.privateIp;
    if (!privateIp || !node.desired || !node.status) continue;

    const statusByKey = new Map(
      node.status.services.map((s) => [serviceKey(s.project, s.service), s]),
    );

    for (const ds of node.desired.services) {
      if (!ds.ingress || ds.ingress.edge !== edgeNodeId) continue;
      const st = statusByKey.get(serviceKey(ds.project, ds.service));
      if (!st) continue;

      for (const replica of st.replicas) {
        if (replica.state === "running" && replica.healthy && replica.hostPort != null) {
          const list = result.get(ds.ingress.domain) ?? [];
          list.push({ domain: ds.ingress.domain, privateIp, hostPort: replica.hostPort });
          result.set(ds.ingress.domain, list);
        }
      }
    }
  }

  return result;
}

/** Build Caddy upstreams from app-published routing shards (edge agent read path). */
export function buildEdgeBackendsFromShards(shards: UpstreamShard[]): Map<string, EdgeBackend[]> {
  const result = new Map<string, EdgeBackend[]>();

  for (const shard of shards) {
    for (const b of shard.backends) {
      const list = result.get(b.domain) ?? [];
      list.push({ domain: b.domain, privateIp: shard.privateIp, hostPort: b.hostPort });
      result.set(b.domain, list);
    }
  }

  return result;
}

/** First health path published for a domain across shards (for Caddy active checks). */
export function edgeHealthPathForDomain(
  shards: UpstreamShard[],
  domain: string,
): string | undefined {
  for (const shard of shards) {
    for (const b of shard.backends) {
      if (b.domain === domain && b.healthPath) return b.healthPath;
    }
  }
  return undefined;
}
